reference/security.rst

```diff
@@ -0,0 +1,291 @@
+
+Because JAAS is a server side app, the access token and refresh token are stored
+by the back-end application and a session cookie is issued to the browser as described in 
+`Browser Cookies`_.
+
+To protect against `CSRF attacks <https://auth0.com/docs/secure/attack-protection/state-parameters>`__
+the back-end application issues a random nonce used in the ``state`` parameter of the OAuth
+flow. This prevents a malicious attacker from forging a request to login as another user.
+
+Device Code Flow
+^^^^^^^^^^^^^^^^
+
+When performing login via the device flow, the Juju CLI will:
+
+1. Request from JAAS a URL and random code which the user can use to login.
+2. The CLI will then wait for a response from JAAS.
+3. JAAS polls the identity server, which in turn is waiting for the user to complete their login.
+
+A diagram depicting this login flow can be found 
+`here <https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow>`__.
+
+This process does not rely on browser redirects and so is not susceptible to traditional browser vulnerabilities.
+Both the access token and refresh token are obtained and stored by the back-end server
+and the CLI application is issued with a session token as described in `CLI Based Sessions`_.
+
+Client Credential Flow
+^^^^^^^^^^^^^^^^^^^^^^
+
+When there is a need to perform login by a machine rather than a physical user, OAuth handles this through
+the use of a client credential flow.
+
+A diagram depicting this login flow can be found 
+`here <https://auth0.com/docs/get-started/authentication-and-authorization-flow/client-credentials-flow>`__.
+
+This process is reserved for scenarios where machine to machine authentication is required. A good example
+of this is includes the use of the Juju Terraform Provider where the client-credential flow is employed.
+
+In the link above, the client application communicates with the identity provider to retrieve an access token
+that is then used with the API server. This design is shifted in JAAS. The client application instead sends
+its application credentials to JAAS which then forwards the information to the identity provider. JAAS
+effectively acts as a proxy between the client application and the identity provider.
+
+This scheme simplifies authentication for client applications but is only possible since JAAS is a trusted
+application in the system.
+
+Macaroons & Offer Authentication
+--------------------------------
+
+Macaroons are a tool for decentralised authentication similar to JSON Web Tokens.
+The `Go Macaroon package <https://pkg.go.dev/gopkg.in/macaroon.v2@v2.1.0>`__ is used by JAAS and has more
+details on the low-level operations that Macaroons are capable of.
+
+Macaroons are used by Juju for various purposes but in JAAS their primary purpose is for authorising 
+cross-model relations between controllers. When two Juju controllers (that are connected to JAAS) 
```

Comment:
Shouldn't this be cross-controller?

Reply:
Technically they're still called cross-model relations, they just happen to be done across two controllers.